File: AWSinvCleanup.m
%% AWS Inventory Cleanup
% 
% Used to import, remove any quotations within th inventory, remove any
% replicated key data, and then export as a CSV with the necessary
% information. Some of this is necessary after curating a dataset inventory
% on AWS Athena.
%
% This will not work if you use "versioning" on AWS. A new variable will need
% to be included in the import section and be passed on to the export
% table.
%
% Made by Charles Ian O'Leary, 20220914

function [outputCSV] = AWSinvCleanup(inputCSV)
%% Import the inventory CSV
% Set up the Import Options and import the data
opts = delimitedTextImportOptions("NumVariables", 3);

% Specify range and delimiter
opts.DataLines = [2, Inf];
opts.Delimiter = ",";

% Specify column names and types
opts.VariableNamesLine = 1;
opts.VariableTypes = ["char", "char", "char"];

% Specify file level properties
opts.ExtraColumnsRule = "ignore";
opts.EmptyLineRule = "read";

% Import the data
inputT = readtable(inputCSV, opts, 'ReadVariableNames', true);

%% Convert to output type
inputC = table2cell(inputT);
numIdx = cellfun(@(x) ~isnan(str2double(x)), inputC);
inputC(numIdx) = cellfun(@(x) {str2double(x)}, inputC(numIdx));


%% Clear temporary variables
 clear opts
 clear numIdx
%% Remove potential extra "
% AWS Athena like to do this when working with strings
if iscellstr(inputC)
    cleanedInv = strrep(inputC(:,:), '"','');
else
    cleanedInv = inputC;
end

%% Remove any duplicate "key" values. 
% This can be a result of loading multiple inventories into AWS Athena on accident
[~,idx]=unique(  strcat(cleanedInv(:,1),cleanedInv(:,2),cleanedInv(:,3)) , 'rows');
if length(idx) > 1
    cleanedInvWD=cleanedInv(idx,:);
end
%% Create a Table with useful Information
% AWS only wants bucket and key values. 
cleanedInvT = cell2table(cleanedInvWD(:,1:2), 'VariableNames',{'bucket' 'key'});
% Write the final CSV with 'cleaned' before input name
outputCSV = strcat('cleaned', inputCSV);
writetable(cleanedInvT,outputCSV, 'WriteVariableNames',false);

File: AWSinvCleanup.py
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 27 12:32:39 2022

 Used to import, remove any quotations within th inventory, remove any
 replicated key data, and then export as a CSV. Some of this is necessary after 
 selecting a dataset inventory on AWS Athena. Also gives size as GB as output.
 This will not work if you use versioning on AWS. A new variable will need
 to be included in the import section, and a column will need to be
 included in the export table.

@author: cioleary
"""

# Define the function. It requires an input CSV to be passed to it.
def AWSinvCleanup(inputCSV):
    #%% Import necessary modules and the CSV
    import pandas as pd
    import numpy as np
    # Import the CSV as a pandas dataframe
    inputDF = pd.read_csv(inputCSV)
    #%% Remove any "
    # These can appear due to how AWS Athena handles strings
    cleanedDF = inputDF.applymap(lambda x: x.replace('"', '')) 
    #%% Create index 
    # Find any instances of duplicated key values
    idx = cleanedDF.duplicated(subset=['key'])
    # Invert the index. This is due to the symantics of .duplicated 
    idx = np.invert(idx)
    # Reduce the DF based on the indexing
    cleanddDF = cleanedDF[idx]
    #%% If size variable exists, calculate total size of dataset, then drop variable from output CSV
    if 'size' in cleanddDF.columns:  
        size = cleanddDF['size'].astype(str).astype(int)
        sizeGB = size.sum() / (1024 ** 3)
        cleanRedDF = cleanddDF.drop(['size'], axis=1)
    else: cleanRedDF = cleanddDF
    #%% Write the new pandas dataframe to CSV
    # Give the CSV a "cleaned" name
    outputCSV = "cleaned" + inputCSV
    # Write the CSV uning the new name, without a header
    cleanRedDF.to_csv(outputCSV,header=False,index=False)
    # Give the size of the dataset in Gigabytes as an output
    print("Size in GB: " + str(sizeGB))
    return(sizeGB)
